src/pyk/kcfg/explore.py

```
@@ -178,6 +189,73 @@ def cterm_implies(
+                    lhs = extract_lhs(_term)
+                    rhs = extract_rhs(_term)
+                    if lhs == rhs:
+                        return KApply(_term.label, [abstract_term_safely(lhs, base_name=_term.label.name)])
+                return _term
+
+            return bottom_up(_no_cell_rewrite_to_dots, term)
+
+        def _is_cell_subst(csubst: KInner) -> bool:
+            if type(csubst) is KApply and csubst.label.name == '_==K_':
+                csubst_arg = csubst.args[0]
+                if type(csubst_arg) is KVariable and csubst_arg.name.endswith('_CELL'):
+                    return True
+            return False
+
+        def _is_negative_cell_subst(constraint: KInner) -> bool:
+            constraint_bool = ml_pred_to_bool(constraint)
+            if type(constraint_bool) is KApply and constraint_bool.label.name == 'notBool_':
+                negative_constraint = constraint_bool.args[0]
+                if type(negative_constraint) is KApply and negative_constraint.label.name == '_andBool_':
+                    constraints = flatten_label('_andBool_', negative_constraint)
+                    cell_constraints = list(filter(_is_cell_subst, constraints))
+                    if len(cell_constraints) > 0:
+                        return True
+            return False
+
+        config_match = self.cterm_implies(CTerm.from_kast(abstract.config), CTerm.from_kast(concrete.config))
+        if config_match is None:
+            failing_cells = []
+            for cell in concrete.cells:
+                concrete_cell = concrete.cell(cell)
+                abstract_cell = abstract.cell(cell)
+                cell_match = abstract_cell.match(concrete_cell)
+                if cell_match is None:
+                    failing_cell = push_down_rewrites(KRewrite(concrete_cell, abstract_cell))
+                    failing_cell = no_cell_rewrite_to_dots(failing_cell)
+                    failing_cells.append((cell, failing_cell))
```

```suggestion
        config_match = self.cterm_implies(CTerm.from_kast(abstract.config), CTerm.from_kast(concrete.config))
        if config_match is None:
            failing_cells = []
            curr_cell_match = Subst({})
            for cell in concrete.cells:
                concrete_cell = concrete.cell(cell)
                abstract_cell = abstract.cell(cell)
                cell_match = abstract_cell.match(concrete_cell)
                if cell_match is not None:
                    _curr_cell_match = curr_cell_match.union(cell_match)
                    if _curr_cell_match is not None:
                        curr_cell_match = _curr_cell_match
                        continue
                failing_cell = push_down_rewrites(KRewrite(concrete_cell, abstract_cell))
                failing_cell = no_cell_rewrite_to_dots(failing_cell)
                failing_cells.append((cell, failing_cell))
```

---

Perhaps we could accumulate the substitution this way?
It would be good to make sure that with an accumulated substitution, where it succeeds on each cell individually, but fails on some certain specific cells, that the failure is reported correctly.

For example:

```
<k> X </k> <state> 3 |-> X </state>

<k> 5 </k> <state> 3 |-> 6 </state>
```

This test should report a failure on the `<state>` cell likely, because the variable `X` is already bound to `5` by the `<k>` cell.

---

Added the accumulation code and the test. It seems to work (reports only the state cell as failure).